### DISCUSSION TITLE
Help getting reference highlighting working

### DISCUSSION THREAD

**Discussion by ahelwer** [nvim-treesitter/nvim-treesitter#1667] [Q&A]
I've been writing highlights for [my tree-sitter grammar](https://github.com/tlaplus-community/tree-sitter-tlaplus/), just basic stuff like `(rulename) @capturename`. All is well. Now I'm trying to get definitions & references working and am running into trouble. Consider a simple TLA+ module like:
```tla
---- MODULE Test ----
CONSTANT Foo
op == Foo
====
```
which declares a global constant, `Foo`, then a definition containing a reference to that constant. I would like the reference to be highlighted the same as the constant. The tree-sitter parse tree for this module is:
```
(source_file
  (module (single_line) (identifier) (single_line)
    (unit (constant_declaration (identifier)))     
    (unit (operator_definition (identifier) (def_eq) (identifier_ref)))
  (double_line))
)
```
I can highlight `Foo` as a constant with the query `(constant_declaration (identifier) @constant)` in my highlights.scm file. I've tried the following in my locals.scm file:
```
(module) @scope
(constant_declaration
  ((identifier) @definition.enum)
  (set! "definition.enum.scope" "global")
)
(identifier_ref) @reference
```
and have verified with `:checkhealth` that my language supports both highlights and locals, but the reference remains unhighlighted:
![image](https://user-images.githubusercontent.com/718765/128197339-3957fcf4-a3cd-47b0-8d16-baf993c47020.png)
Am I doing something obviously wrong?

**Comment by vigoux** [comment]
To actually use what the locals give you, you will have to use the `is?` predicate.

Unfortunately, this predicate is really bad in performances, and I would suggest not to use it for now.

I'll try to get back to this once the locals computation is more efficient.

**Reply by theHamsta** [reply]
I waited with my draft so long that in the mean time parent references (`.parent()`) got expensive :-( So we probably have to walk the changed tree once to get our own parent references.

It would be really cool to use [salsa](https://docs.rs/salsa/0.16.1/salsa/) for such recomputations caching problems and queries that query for the relationship between tree-sitter objects. It is used by rust-analyzer. But unfortunatelly we have a Lua project in the first place :sweat_smile:

**Reply by ahelwer** [reply]
@vigoux @theHamsta what are the `locals.scm` queries currently used for if not this type of reference highlighting? Do they implement go-to-definition/reference finding functionality? Also, do you have any documentation on the `is?` predicate so I can test whether it produces acceptable performance for highlighting? I cannot find it being used in any of the existing .scm files.

**Reply by vigoux** [reply]
For now they are not really used for perormance reason. 
We are trying to solve that, and we will keep you up to date

**Reply by muhmud** [reply]
I wish this was clearer from the README; just burned a number of hours on this thinking I was doing something wrong.

**Reply by ahelwer** [reply]
@vigoux @theHamsta I'll be [giving a demo](https://conf.tlapl.us/2021/andrewhelwer/) of this in a couple days, and it would be nice to just show this feature working - I don't really care about performance, it'll be a small file, I'd just like to give people a glimpse of what's possible & coming in the future. The changes will only live on my laptop and won't be merged. Is it possible to hack something like that together? If so, how?

**Reply by theHamsta** [reply]
Isn't the `is?` predicate working for you? It should still be in nvim-treesitter master. You can basically ask what kind of symbol it's definition is. E.g. if you use an indentifier and there is a function with this name in scope you can apply a different highlight.

**Reply by ahelwer** [reply]
@theHamsta do you have an example of how to use that?

**Reply by theHamsta** [reply]
This used to be faster until Tree-Sitter made calling `:parent()` slow. https://github.com/nvim-treesitter/nvim-treesitter/pull/1120/files it uses `is-fast?` which has the same usage as the normal `is?` in master

**Reply by theHamsta** [reply]
It's possible that the original `is?` was used without the `definition.` prefix.  Safest would be to look into the actual commits that used this. You can also use Atom which has this feature properly implemented. But ours should work also apart from performance considerations.

**Reply by theHamsta** [reply]
https://github.com/nvim-treesitter/nvim-treesitter/pull/296/files

So apparently without the `definion.`

```
((identifier) @parameter.reference
 (#is? @parameter.reference parameter))
```

You probably need to add quotes around "parameter" with the new Tree-Sitter versions.it will look up the definition of the found node and apply the highlight only if the definition is a parameter as defined as `@definition.parameter` in locals.scm. `is? @foo "XXX"` will check for `@defijtiom.XXX`.

**Reply by ahelwer** [reply]
@theHamsta unfortunately I can't get it to work. Given this file:
```tla
---- MODULE Test ----
CONSTANT Foo
op == Foo
====
```
with this parse tree:
```
(source_file [0, 0] - [4, 0]
  (module [0, 0] - [3, 4]
    (header_line [0, 0] - [0, 4])
    name: (identifier [0, 12] - [0, 16])
    (header_line [0, 17] - [0, 21])
    (constant_declaration [1, 0] - [1, 12]
      (identifier [1, 9] - [1, 12]))
    (operator_definition [2, 0] - [2, 9]
      name: (identifier [2, 0] - [2, 2])
      (def_eq [2, 3] - [2, 5])
      definition: (identifier_ref [2, 6] - [2, 9]))
    (double_line [3, 0] - [3, 4])))
```
I have this in my locals.scm:
```
(constant_declaration (identifier) @definition.parameter
  (set! "definition.parameter.scope" "global")
)
```
and this in my highlights.scm (among other highlight queries):
```
((identifier_ref) @parameter.reference
  (#is? @parameter.reference "parameter")
)
```
(also tried it without quoting parameter)
unfortunately the Foo identifier_ref node remains un-highlighted. Would greatly appreciate any debugging strategies you can provide!

**Reply by ahelwer** [reply]
Update: there's no urgency here as the demo was today; it went just fine without reference highlighting! nvim-treesitter is impressive enough as it is!

**Reply by susliko** [reply]
@vigoux @theHamsta 
Is there any progress on making scope-aware variable highlighting feature accessible from Neovim? 👀

**Reply by theHamsta** [reply]
No, there's no progress on that. As I said, it's already working but might cause lags on large files. Feel free on how you would implement https://github.com/nvim-treesitter/nvim-treesitter/pull/1120 in a world where calling `:parent()` is expensive (e.g. parent cache).
